### PR TITLE
KTOR-6525 Migrate to JUnit5

### PR DESCRIPTION
**Pull request by bjhham**
**Subsystem**
Test Suite

**Motivation**
KTOR-6525 Migrate to JUnit5

**Solution**
- Parameterized tests now require arguments in the method.
- Test execution requires Java 11 now, so `javaLauncher` is set for `jvmTest` task
- CoroutinesTimeout is now an annotation that can be applied at the class or function level
- Used `kotlin.test` where possible
- Many other little quirks

### CODE REVIEW

**Comment by bjhham** [comment]
I'll still need to mess around with the timeouts a bit, all the tests work locally though.

File: ktor-client/ktor-client-cio/jvm/test/io/ktor/client/engine/cio/CIORequestTest.kt
Comment (e5l) [inline ktor-client/ktor-client-cio/jvm/test/io/ktor/client/engine/cio/CIORequestTest.kt:54]: Could we use class-level annotation instead?
https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-debug/kotlinx.coroutines.debug.junit5/-coroutines-timeout/

File: ktor-server/ktor-server-test-suites/jvm/src/io/ktor/server/testing/suites/ConfigTestSuite.kt
Comment (e5l) [inline ktor-server/ktor-server-test-suites/jvm/src/io/ktor/server/testing/suites/ConfigTestSuite.kt:59]: Could you tell me why raw assert instead of assertTrue is used?

File: ktor-server/ktor-server-test-host/jvm/src/io/ktor/server/testing/BaseTestJvm.kt
Comment (e5l) [inline ktor-server/ktor-server-test-host/jvm/src/io/ktor/server/testing/BaseTestJvm.kt:19]: Should we keep `ktor-server-test-host` using junit4 and make a separate module for junit5? Let's discuss

File: ktor-server/ktor-server-jetty-jakarta/build.gradle.kts
Comment (e5l) [inline ktor-server/ktor-server-jetty-jakarta/build.gradle.kts:7]: could you tell me why do we need test dependency in the main module?

File: ktor-server/ktor-server-plugins/ktor-server-metrics/jvm/test/io/ktor/server/metrics/dropwizard/DropwizardMetricsTests.kt
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-metrics/jvm/test/io/ktor/server/metrics/dropwizard/DropwizardMetricsTests.kt:98]: we could use `assertAll` from junit 5 or make our own extension

**Review by e5l — CHANGES_REQUESTED** [review]
Thanks for the PR, it's mostly good. Please check the dependencies and some minor stuff.

File: ktor-shared/ktor-junit/build.gradle.kts
Comment (bjhham) [reply ktor-shared/ktor-junit/build.gradle.kts:5]: Shouldn't be, I just missed the configuration for avoiding the publish.

How is that done, is it here in `build.gradle.kts`?

```kotlin
extra["skipPublish"] = mutableListOf<String>()
```

File: ktor-server/ktor-server-test-suites/jvm/src/io/ktor/server/testing/suites/ConfigTestSuite.kt
Comment (bjhham) [reply ktor-server/ktor-server-test-suites/jvm/src/io/ktor/server/testing/suites/ConfigTestSuite.kt:59]: Oh I meant to replace with `kotlin.test` which just has the arguments swapped, I'll fix.

File: ktor-server/ktor-server-jetty-jakarta/build.gradle.kts
Comment (bjhham) [reply ktor-server/ktor-server-jetty-jakarta/build.gradle.kts:7]: Good catch, I must have got my build scripts mixed up.

File: ktor-server/ktor-server-test-host/jvm/src/io/ktor/server/testing/BaseTestJvm.kt
Comment (bjhham) [reply ktor-server/ktor-server-test-host/jvm/src/io/ktor/server/testing/BaseTestJvm.kt:19]: The problem here is that it's used in other projects, so we'd have a swath of tests using junit 4.  The only big difference here is that we're going from 60 seconds timeout for each test to 5 minutes total because of how the annotations work.  You can override the annotation on the child classes and methods if another timeout is needed, though it doesn't seem to be a problem atm...

File: ktor-client/ktor-client-cio/jvm/test/io/ktor/client/engine/cio/CIORequestTest.kt
Comment (bjhham) [reply ktor-client/ktor-client-cio/jvm/test/io/ktor/client/engine/cio/CIORequestTest.kt:54]: Yeah I originally had moved it to the class annotation but it was timing out because it applies for the execution of the whole collection of tests.  I can just sum them up and use the one annotation.

File: ktor-server/ktor-server-test-host/jvm/src/io/ktor/server/testing/BaseTestJvm.kt
Comment (e5l) [reply ktor-server/ktor-server-test-host/jvm/src/io/ktor/server/testing/BaseTestJvm.kt:19]: lgtm

File: ktor-shared/ktor-junit/build.gradle.kts
Comment (bjhham) [reply ktor-shared/ktor-junit/build.gradle.kts:5]: We've got some other internal test projects that aren't included there...  Is the idea that external developers can use those projects to test new engines etc?

**Review by e5l — APPROVED** [review]

**Comment by e5l** [comment]
Please wait for CI builds before merging

File: ktor-shared/ktor-junit/build.gradle.kts
Comment (e5l) [reply ktor-shared/ktor-junit/build.gradle.kts:5]: Exactly

**Review by rsinukov — APPROVED** [review]

**Pull request merged**
